### PR TITLE
Fix issue with assigning an invalid hexadecimal value to float type

### LINKED ISSUE #32750: Bad Sad when assigning hex indicator to float type descriptor
**Description:**
$title

**Steps to reproduce:**
```
public function main() {
    float b = 0x;
    io:println(b);

    b = 0X;
    io:println(b);   
}
```

> [2021-09-15 11:35:06,253] SEVERE {b7a.log.crash} - class java.lang.String cannot be cast to class java.lang.Long (java.lang.String and java.lang.Long are in module java.base of loader 'bootstrap') 
> java.lang.ClassCastException: class java.lang.String cannot be cast to class java.lang.Long (java.lang.String and java.lang.Long are in module java.base of loader 'bootstrap')
>         at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.setLiteralValueAndGetType(TypeChecker.java:473)
>         at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.visit(TypeChecker.java:413)
>         at org.wso2.ballerinalang.compiler.tree.expressions.BLangLiteral.accept(BLangLiteral.java:50)
>         at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.checkExpr(TypeChecker.java:376)
>         at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.checkExpr(TypeChecker.java:355)
>         at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:871)
>         at org.wso2.ballerinalang.compiler.tree.BLangSimpleVariable.accept(BLangSimpleVariable.java:53)
>         at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3497)
>         at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3465)
>         at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeDef(SemanticAnalyzer.java:3457)
>         at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:1562)
>         at org.wso2.ballerinalang.compiler.tree.statements.BLangSimpleVariableDef.accept(BLangSimpleVariableDef.java:46)
>         at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3497)
>         at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3465)
>         at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeStmt(SemanticAnalyzer.java:3461)
>         at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:470)
>         at org.wso2.ballerinalang.compiler.tree.BLangBlockFunctionBody.accept(BLangBlockFunctionBody.java:57)
>         at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3497)
>         at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:440)
>         at org.wso2.ballerinalang.compiler.tree.BLangFunction.accept(BLangFunction.java:73)
>         at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3497)
>         at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:3465)
>         at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeDef(SemanticAnalyzer.java:3457)
>         at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:327)
>         at org.wso2.ballerinalang.compiler.tree.BLangPackage.accept(BLangPackage.java:165)
>         at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyze(SemanticAnalyzer.java:292)
>         at io.ballerina.projects.internal.CompilerPhaseRunner.typeCheck(CompilerPhaseRunner.java:204)
>         at io.ballerina.projects.internal.CompilerPhaseRunner.performTypeCheckPhases(CompilerPhaseRunner.java:112)
>         at io.ballerina.projects.ModuleContext.compileInternal(ModuleContext.java:383)
>         at io.ballerina.projects.ModuleCompilationState$1.compile(ModuleCompilationState.java:45)
>         at io.ballerina.projects.ModuleContext.compile(ModuleContext.java:329)
>         at io.ballerina.projects.PackageCompilation.compileModulesInternal(PackageCompilation.java:174)
>         at io.ballerina.projects.PackageCompilation.compileModules(PackageCompilation.java:166)
>         at io.ballerina.projects.PackageCompilation.from(PackageCompilation.java:95)
>         at io.ballerina.projects.PackageContext.getPackageCompilation(PackageContext.java:206)
>         at io.ballerina.projects.Package.getCompilation(Package.java:135)
>         at io.ballerina.cli.task.CompileTask.execute(CompileTask.java:64)
>         at io.ballerina.cli.TaskExecutor.executeTasks(TaskExecutor.java:40)
>         at io.ballerina.cli.cmd.RunCommand.execute(RunCommand.java:163)
>         at java.base/java.util.Optional.ifPresent(Optional.java:183)
>         at io.ballerina.cli.launcher.Main.main(Main.java:58)

### LINKED ISSUE #32817: Bad Sad when assigning 0xFFFFFFFFFFFFFFFF to Float type
**Description:**
compiler crashes when attempting to assign a hex number that exceeds 15 characters, when it includes all ( a .. f | A .. F) or a combination of (0..9 | a .. f | A .. F)

**Steps to reproduce:**
```
public function main() {
    float d = 0xFFFFFFFFFFFFFFFF;
    io:println(d);

    d = 0xabc435de769FEAB0;
    io:println(d);

    d = 0xaaaaaaaaaaaaaaa0;
    io:println(d);

    d = 0xAAAAAAAAAAAAAAA0;
    io:println(d);
}
```

### PR DESCRIPTION
**Pull request by lasinicl**
## Purpose
$title

Fixes #32817
Fixes #32750

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/resources/test-src/types/float/float_type_negative.bal [hunk 1]
```diff
@@ -0,0 +1,33 @@
+function testInvalidHexaDecimalWithFloatType() {
```
Comment (gimantha) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/float/float_type_negative.bal:1]: Shall check this case too?

```ballerina
type FOO 0xFFFFFFFFFFFFFFFF;
function test() {
    FOO dd = 0xFFFFFFFFFFFFFFFF;
}
```

 it is throwing a Parsing exception.
Also lets write a test case for ReferredTypes

e.g. 
```ballerina
type BOO float;
BOO d = 0xFFFFFFFFFFFFFFFF
```
Second example is giving the expected error.

**Comment by lochana-chathura** [comment]
Fixes #32750 as well. Let's add a test case.

File: tests/jballerina-unit-test/src/test/resources/test-src/types/float/float_type_negative.bal (hunk 1 above)
Comment (lochana-chathura) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/float/float_type_negative.bal:1]: Seems the parsing exception is the same as #32839.

Comment (gimantha) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/float/float_type_negative.bal:1]: Lets merge this fix then, let @rdulmina add the testcase in his PR

Comment (lasinicl) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/float/float_type_negative.bal:1]: As per offline discussion, will not be fixing the above issue with this PR.

File: tests/jballerina-unit-test/src/test/resources/test-src/types/float/float_type_negative.bal [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+float|decimal y3 = 0xaaaaaaaaaaaaaaa0;
+
+float|decimal y4 = 0xAAAAAAAAAAAAAAA0;
+
+type FloatType float;
+FloatType u1 = 0xFFFFFFFFFFFFFFFF;
+
+type FloatType2 float|decimal;
+FloatType2 u2 = 0xFFFFFFFFFFFFFFFF;
```
Comment (lochana-chathura) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/float/float_type_negative.bal:39]: correct formatting is
```suggestion
type FloatType float;

FloatType u1 = 0xFFFFFFFFFFFFFFFF;

type FloatType2 float|decimal;

FloatType2 u2 = 0xFFFFFFFFFFFFFFFF;
```

**Review by lochana-chathura — APPROVED** [review]

File: tests/jballerina-unit-test/src/test/resources/test-src/types/float/float_type_negative.bal (hunk 1 above)
Comment (rdulmina) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/float/float_type_negative.bal:1]: This is partaly fixed in [b0c5a46](https://github.com/ballerina-platform/ballerina-lang/pull/35646/commits/b0c5a46b37e9ced6c47ba77577e7112ac0cc3691) now failing due to  #35695

**Pull request merged**